packages/vscode-extension/src/webview/components/UrlSelect.tsx

```diff
@@ -22,59 +60,208 @@ interface UrlSelectProps {
+// URLs in the dropdown. This can be implemented in the future if needed.
 function UrlSelect({ onValueChange, recentItems, items, value, disabled }: UrlSelectProps) {
-  // We use two lists for URL selection: one with recently used URLs and another
-  // with all available URLs. Since recentItems is a subset of items, each recentItems's
-  // value is prefixed to differentiate their origins when presented in the Select
-  // component. This prefix is stripped off when the selected value is passed back
-  // through onValueChange.
-
-  const handleValueChange = (newSelection: string) => {
-    const stripped = newSelection.replace(/^recent#/, "");
-    onValueChange(stripped);
+  const [isDropdownOpen, setIsDropdownOpen] = React.useState(false);
+  const [inputValue, setInputValue] = React.useState("");
+  const [filteredItems, setFilteredItems] = React.useState<UrlItem[]>([]);
+  const [filteredOutItems, setFilteredOutItems] = React.useState<UrlItem[]>([]);
+
+  const [textfieldWidth, setTextfieldWidth] = React.useState<number>(0);
+  const textfieldRef = useRef<HTMLInputElement>(null);
+
+  const getNameFromId = (id: string) => {
+    const item = items.find((item) => item.id === id);
+    if (!item) {
+      return id;
+    }
+    if (item.name.startsWith("/")) {
+      return item.name;
+    }
+    return item.id;
+  };
+
+  const closeDropdownWithValue = (id: string) => {
+    setInputValue(getNameFromId(id));
+    onValueChange(id);
+    setIsDropdownOpen(false);
+  };
+
+  const navigateBetweenItems = (
+    e: React.KeyboardEvent,
+    prev?: UrlSelectFocusable,
+    next?: UrlSelectFocusable,
+    prevFallback?: UrlSelectFocusable,
+    nextFallback?: UrlSelectFocusable
+  ) => {
+    e.preventDefault();
+    let targetItem = null;
+    let targetItemFallback = null;
+    if (e.key === "ArrowDown") {
+      targetItem = next;
+      targetItemFallback = nextFallback;
+    } else if (e.key === "ArrowUp") {
+      targetItem = prev;
+      targetItemFallback = prevFallback;
+    }
+    if (
+      targetItem &&
+      (targetItem.classList.contains("url-select-item") ||
+        targetItem.classList.contains("url-select-input"))
+    ) {
+      targetItem.focus();
+    } else if (targetItemFallback) {
+      targetItemFallback.focus();
+    } else {
+      textfieldRef.current?.focus();
+    }
   };
 
+  useEffect(() => {
+    setInputValue(getNameFromId(value));
+  }, [value]);
+
+  useEffect(() => {
+    if (!disabled) {
+      const inputValueLowerCase = inputValue.toLowerCase();
+      const filtered = items.filter((item) =>
+        getNameFromId(item.id).toLowerCase().includes(inputValueLowerCase)
+      );
+      setFilteredItems(filtered);
+    }
+  }, [inputValue, items]);
+
+  useEffect(() => {
+    const filteredOut = items.filter(
+      (item) => !filteredItems.some((filteredItem) => filteredItem.id === item.id)
+    );
+    setFilteredOutItems(filteredOut);
+  }, [items, filteredItems]);
```

Comment:
Since the only place that `filteredItems` are set is in the first effect here, I think we could combine these two `useEffect` into a single one.
Then, you could also get the `filteredOutItems` at the same time as you calculate `filteredItems`, instead of checking for each element if it's in the `filteredItems` list (using either a for loop or `lodash`'s [`partition`](https://lodash.com/docs/#partition))
```suggestion
  useEffect(() => {
    if (disabled) {
      setFilteredItems([]);
      setFilteredOutItems(items);
      return;
    }
    
    const inputValueLowerCase = inputValue.toLowerCase();
    const [filtered, filteredOut] = _.partition(items, (item) =>
      getNameFromId(item.id).toLowerCase().includes(inputValueLowerCase)
    );
    setFilteredItems(filtered);
    setFilteredOutItems(filteredOut);
  }, [inputValue, items]);
```
(actually, I'm not sure about the `if (disabled)` behaviour -- I think what you're doing right now is just `return;`, but I'm not sure that's what we want.)